### PR TITLE
Replica promotion in compute_ctl

### PR DESCRIPTION
**Pull request by myrrc**
Add `/promote` method for `compute_ctl` promoting secondary replica to primary,
depends on secondary being prewarmed.
Add `compute-ctl` mode to `test_replica_promotes`, testing happy path only (no corner cases yet)
Add openapi spec for `/promote` and `/lfc` handlers

https://github.com/neondatabase/cloud/issues/19011
Resolves: https://github.com/neondatabase/cloud/issues/29807

### CODE REVIEW

File: compute_tools/src/compute_promote.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        {
+            let state = self.state.lock().unwrap();
+            let mode = &state.pspec.as_ref().unwrap().spec.mode;
+            if *mode != ComputeMode::Replica {
+                bail!("{} is not replica", mode.to_type_str());
+            }
+
+            // we don't need to query Postgres so not self.lfc_prewarm_state()
+            match &state.lfc_prewarm_state {
+                LfcPrewarmState::NotPrewarmed | LfcPrewarmState::Prewarming => {
```
[the remark this answers is not in the document]
Comment (myrrc) [reply compute_tools/src/compute_promote.rs:62]: Why I built logic this way: we want endpoint to be prewarmed when promotion happens to ensure good startup performance. If we failed to prewarm, we tried (at least), but if we didn't, this means we have a bug in the control plane side, so we should notify it with an error.

File: test_runner/fixtures/endpoint/http.py [hunk 2]
```diff
@@ -102,6 +103,18 @@ def offloaded():
 
         wait_until(offloaded)
 
+    def promote(self, safekeepers_lsn: dict[str, Any], disconnect: bool = False):
+        url = f"http://localhost:{self.external_port}/promote"
+        if disconnect:
+            try:  # send first request to start promote and disconnect
+                self.post(url, data=safekeepers_lsn, timeout=0.001)
+            except ReadTimeout:
+                pass  # wait on second request which returns on promotion finish
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply test_runner/fixtures/endpoint/http.py:112]: May be I am missing something...
But if first request just initiates promotion and second - waits end of promotion, why not to have two different requests: i.e. `start_promote` and `wait_promote_completion`?

[the remark this answers is not in the document]
Comment (myrrc) [reply test_runner/fixtures/endpoint/http.py:112]: Normally we would just POST /promote and wait till it finishes. However, we must test the situation that control plane's handler disconnects. This should not abort the promote. When control plane issues another request, it should not trigger second promotion

File: test_runner/regress/test_replica_promotes.py [hunk 3]
```diff
@@ -146,5 +158,80 @@ def test_replica_promotes(neon_simple_env: NeonEnv, pg_version: PgVersion):
...
+        safekeepers = cur.fetchall()[0][0]
+
+    primary.http_client().offload_lfc()
+    primary_endpoint_id = primary.endpoint_id
+    primary.stop(mode="immediate-terminate")
+    assert (lsn := primary.terminate_flush_lsn)
+
+    client = secondary.http_client()
+    client.prewarm_lfc(primary_endpoint_id)
+    safekeepers_lsn = {"safekeepers": safekeepers, "wal_flush_lsn": lsn}
```
[the remark this answers is not in the document]
Comment (myrrc) [reply test_runner/regress/test_replica_promotes.py:222]: That's interesting. Although reloading config and promoting yields "fatal: port not specified" in case there's some garbage input to "safekeepers", the handler still returns 200.
I guess I should check something after `SHOW transaction_read_only`. Maybe writing to some ephemeral table would do

[the remark this answers is not in the document]
Comment (myrrc) [reply test_runner/regress/test_replica_promotes.py]: Update: this test will be moved to other PR with other corner cases

File: compute_tools/src/http/openapi_spec.yaml [hunk 4]
```diff
@@ -83,6 +83,87 @@ paths:
...
+      summary: Promote secondary replica to primary
+      description: ""
+      operationId: promoteReplica
+      requestBody:
+        description: Promote requests data
+        required: true
+        content:
+          application/json:
+            schema:
+                $ref: "#/components/schemas/SafekeepersLsn"
```
Comment (mtyazici) [inline compute_tools/src/http/openapi_spec.yaml:99]: Normally anything we do in terms of updating compute state we do it by sending the spec. What will happen if we only send these things in the body?

Previous spec that was configured will be incorrect. That's a bit confusing.
